### PR TITLE
Implement test discovery

### PR DESCRIPTION
**Pull request by stepancheg**
`testInDir dir ...` lists all directories in `dir` which contains
`run` files, and such directories are considered tests.

This is done to make test addition/maintenance cheaper.

Convert some test directories to `testInDir`, but not all of them
because
* some directories are listed in several test groups
* other directories are have some tests disabled

This is on top of #1733.

### CODE REVIEW

File: tests/Main.idr
Comment (gallais) [inline tests/Main.idr:35]: This looks general enough to go into `Test.Golden` as part of the library.

**Comment by gallais** [comment]
This looks like a good addition to `Test.Golden` but I'm not sure
about using it quite so extensively in `tests/Main.idr`. If we're
probably not going to turn any test off in `typedd`, it's not
necessarily the case in the other directories. And then we'll suddenly
have to revert to manually listing all of the test cases which will
be a massive burden.

**Comment by stepancheg** [comment]
> If we're probably not going to turn any test off in typedd, it's not necessarily the case in the other directories. And then we'll suddenly have to revert to manually listing all of the test cases which will be a massive burden.

I'd prefer we never do manual listing.
* tests for the same features should live in the same directory. Manually listing should not be needed. I kept existing structure to avoid doing too many changes prematurely. E. g. "Tests interactive" should live in a separate directory `test/interactive` rather than in common `tests/idris2` directory and listed explicitly
* additionally we could specify features per-test. E. g. disable a test, or specify codegens per test. And do it inside the test rather than in `tests/Main.idr`

File: tests/Main.idr
Comment (stepancheg) [reply tests/Main.idr:35]: It is generic enough now (except that it calls `exitFailure` which might be not appropriate), but later I'd like to extend this code to discover options based on test content. E. g. read the list of codegens from the `.idr` file in the test.

**Comment by stefan-hoeck** [comment]
Shouldn't an overhaul of the test framework as you are planning it be first discussed as a proposal in the issue track? Please see also the new [contribution guidelines](https://github.com/idris-lang/Idris2/blob/master/CONTRIBUTING.md).

Please consider that every PR opened means it has to be reviewed by one of the few maintainers of Idris2, all of which are researchers with a full schedule, and none of which is doing this as their full time job. Therefore, before writing PRs it might in general be a good idea to either write a bug report or proposal for a new feature first (but again, see the contribution guidelines about this). You might also consider sharing and discussing your ideas with the community on [discord](https://github.com/idris-lang/Idris2/wiki), where you will typically get some response on a simple question like "who would like me to redesign parts of the test framework for the following reasons:" pretty quickly.

**Comment by stepancheg** [comment]
@stefan-hoeck with all due respect, I wish I could do that, but I can't. I have full time job, and it is not development in Idris. When I have a free minute, I work on opensource projects. I have an idea, I implement it quickly and submit a PR and return to my life.

That test framework change was like 15 minutes of work. I just did it, and returned to other non-Idris related things.

Discord is not an option for me: chats are too big distraction. I barely cope with my work chats.

And discussion on the issue tracker would be a waste of _your_ time: because even if I have an idea today, I cannot promise I can return to that idea later, that discussion might be not even implemented.

Moreover, before I started implementing something, it's hard to estimate how feasible is the idea. So proposing an idea even without understanding whether I could implement it or not would be counterproductive.

I can give the example. I had an idea to write meaningful variable names in generated code in RefC codegen (to make debugging easier). I spent two hours trying to implement it, and gave up. If I created an issue in the tracker, likely the idea would be approved, and then it would be awkward situation when I didn't deliver what I announced. And given how busy you guys are, it is unlikely that I could get help implementing it.

I heard may times "show me the code" (not in Idris). Because most ideas are cheap, and bikeshedding of ideas is a waste of time of all participants, but the implementation is what has real value.

I don't mind if PRs are closed because the idea proposed by the PR is considered wrong.

With the feedback in this PR and similar feedback in other PRs, I have a growing feeling that my PRs are net negative for the project. If it is so, just tell me, and I'll stop.

**Comment by stefan-hoeck** [comment]
Thanks for your honest words. I'm afraid what you describe applies to all of us (disclaimer: I am not one of the Idris2 maintainers): Most of us have full time jobs in industry or academia some in fields not even related to computer sciences. Some have family and other interests. But this does not lift the burden of carefully considering and designing a contribution before submitting PRs from us.

Right now, there 29 PRs open, 13 of which are from you. If you consider that reviewing a PR includes trying to figure out what it does and what kind of problem it solves, whether i aligns well with the rest of the project, whether the code is up to project standards, writing answers and suggestions to the contributer, etc. etc, then it is probably safe to assume that this takes at least one or two hours for every non trivial PR from one of the core developers before the PR is merged. That's a lot of work. So I think it would only be fair we contributors do our part, which includes opening issues and writing feature proposals, which forces us to at least thoroughly consider and working out an idea before taking time from the Idris2 core team.

**Comment by stepancheg** [comment]
I don't see how reviewing a PR is more work than reviewing an idea. I see it as the opposite: PR is a clear well defined manifestation of the idea. Discussing the PR in my view is less work than discussing the issue in the tracker. Especially if many ideas from the tracker do not result in PRs.

In any case, my previous comment stands.

**Comment by andrevidela** [comment]
> I don't see how reviewing a PR is more work than reviewing an idea.

reviewing a PR:
- What is this
- Do we want this?
- Why?
- What is this program doing?
- Running the program
- Could this have been done differently?
- Modify the program
- Run the tests agains
- Does this impact performance?
- Benchmark the tests

Reviewing an issue:
- What is this
- Do we want this?
- Why?

A pull request out of nowhere is not a clear manifestation of an idea, in many case it is difficult to retro-engineer what the idea was in the first place, and this is made worse by potential errors or implementation details that show up in the code. I fyou have a full time job in software engineer you probably have a ticketing system with a manager handling priorities and communication with the customer. Basically a way to push the product forward for the least amount of money. Issues are a way to push the project forward for the least amount of effort.

Anyways, PRs are for reviewing code, not for existentialist question about organisation, I encourage you to raise an issue for that.

File: libs/base/System/Directory.idr
Comment (andrevidela) [inline libs/base/System/Directory.idr:110]: I'm slightly confused by this bit. Is this not an infinite loop? I read this as "if the next entry is either the current one or the parent one, loop on `MkDir d` does this work because of the side effect of `prim__dirEntry`?

Comment (stepancheg) [reply libs/base/System/Directory.idr:110]: > does this work because of the side effect of prim__dirEntry?

Yes.

**Comment by stepancheg** [comment]
> Running the program

We have continuous integration for that. And by the way, this intent of this PR is to improve continuous integration so Idris could have better tests, so reviewers wouldn't have to "run the program" or "run the tests".

> in many case it is difficult to retro-engineer what the idea was in the first place

Consider a PR two parts:

* idea explained in the PR description illustrated by attached code, same text would be in the issue description
* follow-up implementation

You can stop after the step one if you think the idea is bad, so you don't need to run the program or whatever. The amount of reviewing work will be the same at worst.

**Comment by andrevidela** [comment]
@stepancheg This will come out dry but I think it's the best way of communicating it:

Here is not a place for arguing, here is a place to review code. If you have concerns about how the project is organized, voice them in the appropriate channels, either: The mailing list, the discord server or in a new issue with the tag `meta`.

**Comment by andrevidela** [comment]
Now back to the code: I gave it a look and it seems like a fine addition, ready to merge @gallais ?

**Comment by gallais** [comment]
> ready to merge @gallais?

As I said in my review, I think `testsInDir` should be moved to
`Test.Golden` to benefit the whole community.

I'm also not too keen on using it so extensively as we don't have
a way to turn tests off. `testsInDir` could perhaps take an additional
argument: a list of tests that have been turned off.

**Comment by andrevidela** [comment]
I feel like taking in argument an extra function`String -> Bool` that tells us if the test needs to run is the way to go. It allows things like 

```
testsInDir (const True) … -- for all tests
testsInDir ("NEW_" `isPrefixOf`) … -- for tests that are just being added while developing 
testsInDir (not . `contains` removed) ... -- only run if the test is not in the set of removed tests
```

**Comment by stepancheg** [comment]
Updated the PR with:
* `testInDir` moved to `Test.Golden`
* `testInDir` now accepts additional `String -> Bool` parameter to filter tests by directory name

File: libs/base/System/Directory.idr
Comment (gallais) [inline libs/base/System/Directory.idr:128]: I think @Russoul has a trick to use `Compose` here so that the monad
used by the `do` block is `io . Either FileError` instead of `io` and
this way you don't need to systematically pack/unpack the `Left/Right`.

Comment (stepancheg) [reply libs/base/System/Directory.idr:128]: Interesting, I didn't know Idris could do that.

I think I got how it can be done explicitly with:

```
(... >>= ...) @{Monad.Compose}
```

but result is not very readable. But I have not found to use `do` notation with that `Compose`. I. e. how to tell `do` notation to use `Monad.Compose` version of `>>=` instead of default `>>=`?

Comment (Russoul) [reply libs/base/System/Directory.idr:128]: Unfortunately, this is the best solution I can think of:
```
namespace IO.Either
  public export
  (>>=) : (IO . Either e) a -> (a -> (IO . Either e) b) -> (IO . Either e) b
  (>>=) = (>>=) @{Monad.Compose {m = IO} {t = Either e}}

  public export
  pure : a -> (IO . Either e) a
  pure = pure @{Applicative.Compose {f = IO} {g = Either e}}

collectDir d = IO.Either.do
  ...

```
It should relieve us from the packing/unpacking, but it has two noticeable flaws:
1) One has to define that namespace before being able to make use of it
2) The publicly exported `(>>=)` and `pure` can potentially result in puzzling errors in any module that depends on `System.Directory` and has ambiguous occurrences of `pure` or `(>>=)`.

The second point is quite concerning. 
So I'd refrain from applying the solution yet.

For some reason I wasn't able to make it work with a locally let-bound specialised `(>>=)`.

Comment (Russoul) [reply libs/base/System/Directory.idr:128]: Oh, it seems that this works though:
```
collectDir d = do
  let (>>=) : (IO . Either e) a -> (a -> (IO . Either e) b) -> (IO . Either e) b
      (>>=) = Prelude.(>>=) @{Monad.Compose {m = IO} {t = Either e}}
  ...   
```

Comment (stepancheg) [reply libs/base/System/Directory.idr:128]: It works! Thanks! This is my version:

```
collectDir : HasIO io => Directory -> io (Either FileError (List String))
collectDir d = liftIO $ do let (>>=) : (IO . Either e) a -> (a -> (IO . Either e) b) -> (IO . Either e) b
                               (>>=) = Prelude.(>>=) @{Monad.Compose {m = IO} {t = Either e}}
                           Just n <- nextDirEntry d
                             | Nothing => pure $ Right []
                           ns <- assert_total $ collectDir d
                           pure $ Right (n :: ns)
```

I'll update the PR later.

Comment (stepancheg) [reply libs/base/System/Directory.idr]: Updated the PR.

**Comment by andrevidela** [comment]
Is this ready to go?

**Review by gallais — APPROVED** [review]

**Pull request merged**
